README.md
# @goa/uuid

[![npm version](https://badge.fury.io/js/%40goa%2Fuuid.svg)](https://npmjs.org/package/@goa/uuid)

`@goa/uuid` is a [fork](https://github.com/kelektiv/node-uuid) of Simple, fast generation of RFC4122 UUIDS written in ES6.

```sh
yarn add @goa/uuid
```

## Table Of Contents

- [Table Of Contents](#table-of-contents)
- [API](#api)
- [`v4(options?: UuidConfig, buffer?: Array|Buffer, offset?: number): string`](#v4options-uuidconfigbuffer-arraybufferoffset-number-string)
  * [`_goa.UuidConfig`](#type-_goauuidconfig)
- [Copyright](#copyright)

<p align="center"><a href="#table-of-contents"><img src="/.documentary/section-breaks/0.svg?sanitize=true"></a></p>

## API

The package is available by importing its default function. Only `v4` is supported.

```js
import v4 from '@goa/uuid'
```

<p align="center"><a href="#table-of-contents"><img src="/.documentary/section-breaks/1.svg?sanitize=true"></a></p>

## `v4(`<br/>&nbsp;&nbsp;`options?: UuidConfig,`<br/>&nbsp;&nbsp;`buffer?: Array|Buffer,`<br/>&nbsp;&nbsp;`offset?: number,`<br/>`): string`

Generate and return a RFC4122 v4 UUID.

__<a name="type-_goauuidconfig">`_goa.UuidConfig`</a>__: Optional uuid state to apply.

|  Name  |                   Type                    |                                  Description                                  |
| ------ | ----------------------------------------- | ----------------------------------------------------------------------------- |
| random | <em>!Array&lt;number&gt;</em>             | Array of 16 numbers (0-255) to use in place of randomly generated values.     |
| rng    | <em>function(): !Array&lt;number&gt;</em> | Random # generator function that returns an Array[16] of byte values (0-255). |

```js
/* alanode example/ */
import uuid from '@goa/uuid/v4'

const res = uuid()
console.log(res)
```
```
66534101-c639-4670-a67d-390258a53fb0
```

<p align="center"><a href="#table-of-contents"><img src="/.documentary/section-breaks/2.svg?sanitize=true"></a></p>

## Copyright

Original work by [Robert Kieffer](https://github.com/kelektiv/node-uuid/blob/master/LICENSE.md) and other contributors.

---

<table>
  <tr>
    <th>
      <a href="https://artd.eco">
        <img src="https://raw.githubusercontent.com/wrote/wrote/master/images/artdeco.png" alt="Art Deco">
      </a>
    </th>
    <th>© <a href="https://artd.eco">Art Deco</a> for <a href="https://idio.cc">Idio</a> 2019</th>
    <th>
      <a href="https://idio.cc">
        <img src="https://avatars3.githubusercontent.com/u/40834161?s=100" width="100" alt="Idio">
      </a>
    </th>
    <th>
      <a href="https://www.technation.sucks" title="Tech Nation Visa">
        <img src="https://raw.githubusercontent.com/artdecoweb/www.technation.sucks/master/anim.gif"
          alt="Tech Nation Visa">
      </a>
    </th>
    <th><a href="https://www.technation.sucks">Tech Nation Visa Sucks</a></th>
  </tr>
</table>

<p align="center"><a href="#table-of-contents"><img src="/.documentary/section-breaks/-1.svg?sanitize=true"></a></p>

CHANGELOG.md
## 29 June 2019

### [1.0.2](https://github.com/idiocc/uuid/compare/v1.0.1...v1.0.2)

- [package] Correct the `main` field.

## 28 June 2019

### [1.0.1](https://github.com/idiocc/uuid/compare/v1.0.0...v1.0.1)

- [package] Rearrange files.

### [1.0.0](https://github.com/idiocc/uuid/compare/v0.0.0-pre...v1.0.0)

- [package] Publish `v1` of the package.

## 27 June 2019

### [1.0.0](https://github.com/idiocc/uuid/compare/v0.0.0-pre...v1.0.0)

- [package] Publish the first version.

### 0.0.0-pre

- Create `@goa/uuid` with _[`My New Package`](https://mnpjs.org)_
- [repository]: `src`, [`test`](https://contexttesting.com), [`documentary`](https://readme.page) & [`types`](https://typedef.page).

test/mask/default.js
import makeTestSuite from '@zoroaster/mask'
import Context from '../context'
import uuid from '../../src'

// export default
makeTestSuite('test/result', {
  async getResults() {
    const res = await uuid({
      text: this.input,
    })
    return res
  },
  context: Context,
})

test/spec/default.js
import { equal, ok } from '@zoroaster/assert'
import Context from '../context'
import uuid from '../../src'

/** @type {Object.<string, (c: Context)>} */
const T = {
  context: Context,
  'is a function'() {
    equal(typeof uuid, 'function')
  },
  async 'calls package without error'() {
    await uuid()
  },
  async 'gets a link to the fixture'({ fixture }) {
    const text = fixture`text.txt`
    const res = await uuid({
      text,
    })
    ok(res, text)
  },
}

export default T

test/context/index.js
import { join } from 'path'
import { debuglog } from 'util'

const LOG = debuglog('@goa/uuid')

/**
 * A testing context for the package.
 */
export default class Context {
  async _init() {
    LOG('init context')
  }
  /**
   * Example method.
   */
  example() {
    return 'OK'
  }
  /**
   * A tagged template that returns the relative path to the fixture.
   * @param {string} file
   * @example
   * fixture`input.txt` // -> test/fixture/input.txt
   */
  fixture(file) {
    const f = file.raw[0]
    return join('test/fixture', f)
  }
  async _destroy() {
    LOG('destroy context')
  }
}

src/index.js
import rng from './rng'
import bytesToUuid from './bytes-to-uuid'

/**
 * Generate and return a RFC4122 v4 UUID.
 * @param {_goa.UuidConfig|string} options Options, or "binary" string.
 * @param {Array|Buffer} [buffer] Array or buffer where UUID bytes are to be written.

 * @param {number} [offset] Starting index in `buffer` at which to begin writing.
 */
export default function v4(options = {}, buffer = null, offset = 0) {
  var i = buffer && offset

  if (typeof options == 'string') {
    buffer = options == 'binary' ? new Array(16) : null
    options = null
  }
  const { random, rng: rngFn = rng } = /** @type {_goa.UuidConfig} */ (options)

  const rnds = random || rngFn()

  // Per 4.4, set bits for version and `clock_seq_hi_and_reserved`
  rnds[6] = (rnds[6] & 0x0f) | 0x40
  rnds[8] = (rnds[8] & 0x3f) | 0x80

  // Copy bytes to buffer, if provided
  if (buffer) {
    for (var ii = 0; ii < 16; ++ii) {
      buffer[i + ii] = rnds[ii]
    }
  }

  return buffer || bytesToUuid(rnds)
}

/**
 * @suppress {nonStandardJsDocs}
 * @typedef {import('./types').UuidConfig} _goa.UuidConfig
 */

v4.js
import rng from './src/rng'
import bytesToUuid from './src/bytes-to-uuid'

/**
 * Generate and return a RFC4122 v4 UUID.
 * @param {_goa.UuidConfig|string} options Options, or "binary" string.
 * @param {Array|Buffer} [buffer] Array or buffer where UUID bytes are to be written.

 * @param {number} [offset] Starting index in `buffer` at which to begin writing.
 */
export default function v4(options = {}, buffer = null, offset = 0) {
  var i = buffer && offset

  if (typeof options == 'string') {
    buffer = options == 'binary' ? new Array(16) : null
    options = null
  }
  const { random, rng: rngFn = rng } = /** @type {_goa.UuidConfig} */ (options)

  const rnds = random || rngFn()

  // Per 4.4, set bits for version and `clock_seq_hi_and_reserved`
  rnds[6] = (rnds[6] & 0x0f) | 0x40
  rnds[8] = (rnds[8] & 0x3f) | 0x80

  // Copy bytes to buffer, if provided
  if (buffer) {
    for (var ii = 0; ii < 16; ++ii) {
      buffer[i + ii] = rnds[ii]
    }
  }

  return buffer || bytesToUuid(rnds)
}

/**
 * @suppress {nonStandardJsDocs}
 * @typedef {import('./types').UuidConfig} _goa.UuidConfig
 */